nullaway/src/main/java/com/uber/nullaway/handlers/ContractHandler.java

```diff
@@ -0,0 +1,192 @@
+import com.uber.nullaway.Nullness;
+import com.uber.nullaway.dataflow.AccessPath;
+import com.uber.nullaway.dataflow.AccessPathNullnessPropagation;
+import java.util.Map;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.ExecutableElement;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+
+/**
+ * This Handler parses the jetbrains @Contract annotation and honors the nullness spec defined there
+ * on a best effort basis.
+ *
+ * <p>Currently, we can only reason about cases where the contract specifies that the return value
+ * of the method depends on the nullness value of a single argument. This means we can reason about
+ * rules like the following:
+ *
+ * <ul>
+ *   <li>@Contract("null -> true")
+ *   <li>@Contract("_, null, _ -> false")
+ *   <li>@Contract("!null, _ -> false; null, _ -> true")
+ *   <li>@Contract("!null -> !null")
+ * </ul>
+ *
+ * In the last case, nullness will be propagated iff the nullness of the argument is already known
+ * at invocation.
+ *
+ * <p>However, when the return depends on multiple arguments, this handler usually ignores the rule,
+ * since it is not clear which of the values in question are null or not. For example,
+ * for @Contract("null, null -> true") we know nothing when the method returns true (because truth
+ * of the consequent doesn't imply truth of the antecedent), and if it return false, we only know
+ * that at least one of the two arguments was non-null, but can't know for sure which one. NullAway
+ * doesn't reason about multiple value conditional nullness constraints in any general way.
+ *
+ * <p>In some cases, this handler can determine that some arguments are already known to be non-null
+ * and reason in terms of the remaining (under-constrained) arguments, to see if the final value of
+ * this method depends on the nullness of a single argument for this callsite, even if the @Contract
+ * clause is given in terms of many. This is not behavior that should be counted on, but it is
+ * sound.
+ */
+public class ContractHandler extends BaseNoOpHandler {
+
+  @Override
+  public NullnessHint onDataflowVisitMethodInvocation(
+      MethodInvocationNode node,
+      Types types,
+      AccessPathNullnessPropagation.SubNodeValues inputs,
+      AccessPathNullnessPropagation.Updates thenUpdates,
+      AccessPathNullnessPropagation.Updates elseUpdates,
+      AccessPathNullnessPropagation.Updates bothUpdates) {
+    Symbol.MethodSymbol callee = ASTHelpers.getSymbol(node.getTree());
+    Preconditions.checkNotNull(callee);
+    // Check to see if this method has an @Contract annotation
+    if (ASTHelpers.hasDirectAnnotationWithSimpleName(callee, "Contract")) {
+      // Found a contract, lets parse it.
+      String[] clauses = getContractFromAnnotation(callee).split(";");
+      for (String clause : clauses) {
+        String[] parts = clause.split("->");
+        assert parts.length == 2;
```

Comment:
Will very bad things happen if `parts.length != 2`?  If so, we should throw an exception here and not rely on asserts being enabled

Reply:
No. Everything past the second "->" will simply be ignored. I guess the common typo might be something like `"!null -> -> !null"`. This would result in the consequent being `->` and the clause being silently ignored. Which means that part of the annotation would have no effect, and we should still be as sound as we were before this patch.

That said, maybe there is value in more systematically validating this clauses. I am more worried about the following assert below that one (`assert antecedent.length == node.getArguments().size()`). Because that one failing means the annotation writer is seriously confused, yet an error in one direct would result in it being interpreted as a valid (probably erroneous) spec, and the other would crash with an index out of bounds error.

I'll look into this tomorrow. Probably worth fixing before landing.